src/main/scala/io/iohk/ethereum/db/storage/ReferenceCountNodeStorage.scala

```diff
@@ -36,99 +36,106 @@ class ReferenceCountNodeStorage(nodeStorage: NodeStorage, pruningOffset: BigInt,
-        .orElse(nodeStorage.get(nodeKey).map(storedNodeFromBytes)) // or db
-        .map(_.decrementReferences(1))
-
-      if (storedNode.isDefined) {
-        val pruneCanditatesUpdated =
-          // if references is 0, mark it as prune candidate for current block tag
-          if (storedNode.get.references == 0) toDeleteInBlock.copy(nodeKeys = nodeKey +: toDeleteInBlock.nodeKeys)
-          else toDeleteInBlock
-        (storedNodes + (nodeKey -> storedNode.get), pruneCanditatesUpdated)
+    val changes = toRemove.foldLeft(upsertChanges) { (storedNodes, nodeKey) =>
+      val maybeStoredNode: Option[(StoredNode, StoredNodeSnapshot)] = storedNodes.get(nodeKey) // get from current changes
+        .orElse(nodeStorage.get(nodeKey).map(storedNodeFromBytes).map(s => s -> StoredNodeSnapshot(nodeKey, Some(s)))) // or db
+
+      if (maybeStoredNode.isDefined) {
+        val (storedNode, snapshot) = maybeStoredNode.get
+        storedNodes + (nodeKey -> (storedNode.decrementReferences(1), snapshot))
       }
-      else acc
+      else storedNodes
     }
 
-    // map stored nodes to bytes in order to save them
-    val toUpsertUpdated = changes._1.map {
-      case (nodeKey: NodeHash, storedNode: StoredNode) => nodeKey -> storedNodeToBytes(storedNode)
-    }.toSeq
+    val (toRemoveUpdated, toUpsertUpdated, snapshots) =
+      changes.foldLeft(Seq.empty[NodeHash], Seq.empty[(NodeHash, NodeEncoded)], Seq.empty[StoredNodeSnapshot]) {
+        case ((removeAcc, upsertAcc, snapshotAcc), (key, (storedNode, theSnapshot))) =>
+          // If no more references, move it to the list to be removed
+          if(storedNode.references == 0) (removeAcc :+ key, upsertAcc, snapshotAcc :+ theSnapshot)
+          else (removeAcc, upsertAcc :+ (key -> storedNodeToBytes(storedNode)), snapshotAcc :+ theSnapshot)
+      }
 
-    val toMarkAsDeleted =
-      // Update prune candidates for current block tag if it references at least one block that should be removed
-      if (changes._2.nodeKeys.nonEmpty) Seq(toPruneInThisBlockKey -> pruneCandiatesToBytes(changes._2))
+    val snapshotToSave = {
+      if(snapshots.nonEmpty) {
+        // If not empty, snapshots will be stored indexed by block number and index
+        val snapshotCountKey = getSnapshotsCountKey(bn)
+        val blockNumberSnapshotsCount: BigInt = nodeStorage.get(snapshotCountKey).map(snapshotsCountFromBytes).getOrElse(0)
+        val snapshotsToSave = snapshots.zipWithIndex.map { case (snapshot, index) =>
+          getSnapshotKey(bn, blockNumberSnapshotsCount + index + 1) -> snapshotToBytes(snapshot)
+        }
+        // Save snapshots and latest snapshot index
+        (snapshotCountKey -> snapshotsCountToBytes(blockNumberSnapshotsCount + snapshotsToSave.size)) +: snapshotsToSave
+      }
       else Nil
-
-    nodeStorage.update(Nil, toUpsertUpdated ++ toMarkAsDeleted)
-
+    }
+    nodeStorage.update(toRemoveUpdated, toUpsertUpdated ++ snapshotToSave)
     this
```

Comment:
generally, i find this function pretty long and having much logic, wdyt about splitting it into few smaller ones?

Reply:
I think it's a good idea, I was thinking to that too.